addon/dom-event-listeners.js

```diff
@@ -0,0 +1,202 @@
+import Ember from 'ember';
+import { run } from '@ember/runloop';
+import { assert } from '@ember/debug';
+import { registerDisposable } from './utils/disposable';
+
+const { WeakMap } = Ember;
+
+const eventListeners = new WeakMap();
+
+const PASSIVE_SUPPORTED = (() => {
+  let ret = false;
+
+  try {
+    let options = Object.defineProperty({}, 'passive', {
+      get() {
+        ret = true;
+      },
+    });
+
+    window.addEventListener('test', null, options);
+  } catch (err) {
+    // intentionally empty
+  }
+  return ret;
+})();
+
+const LISTENER_ITEM_LENGTH = 5;
+const INDEX = {
+  ELEMENT: 0,
+  EVENT_NAME: 1,
+  CALLBACK: 2,
+  ORIGINAL_CALLBACK: 3,
+  OPTIONS: 4,
+};
+
+/**
+   Attaches an event listener that will automatically be removed when the host
+   object is dropped from DOM.
+
+   Example:
+
+   ```js
+   import Component from 'ember-component';
+   import { addEventListener } from 'ember-lifeline';
+
+   export default Component.extend({
+     didInsertElement() {
+       addEventListener(this, '.some-item', 'click', (e) => {
+         console.log('.some-item was clicked');
+       });
+     }
+   });
+   ```
+
+   This can also be used in other ember types like services and controllers. In
+   order to use it there an html element reference must be used instead of a
+   css selector. This way we can be sure the element actually exists when the
+   listener is attached:
+
+   ```js
+   import Service from 'ember-service';
+   import { addEventListener } from 'ember-lifeline';
+
+   export default Service.extend({
+     init() {
+       this._super(...arguments);
+       const el = document.querySelector('.foo');
+       addEventListener(this, el, 'click')
+     }
+   });
+   ```
+
+   @method addEventListener
+   @param { Object } obj the instance to attach the listener for
+   @param { String } selector the DOM selector or element
+   @param { String } _eventName the event name to listen for
+   @param { Function } _callback the callback to run for that event
+   @public
+   */
+export function addEventListener(obj, selector, eventName, _callback, options) {
```

Comment:
Need to find a better names for `(add|remove)EventListener` as these collide with the globals in window.